src/App.js

```diff
@@ -1,32 +1,40 @@
-import React from "react";
+import React, { useState, useEffect } from "react";
 import TodoList from "./TodoList";
 import AddTodoForm from "./AddTodoForm";
 
-const useSemiPersistentState = () => {
-  const [todoList, setTodoList] = React.useState(
-    JSON.parse(localStorage.getItem("savedTodoList")) || []
-  );
-
-  React.useEffect(() => {
-    localStorage.setItem("savedTodoList", JSON.stringify(todoList));
-  }, [todoList]);
-
-  return [todoList, setTodoList];
-};
-
-
 function App() {
-  // Using the custom hook to get and set the todoList state
-  const [todoList, setTodoList] = useSemiPersistentState();
-  
-  function addTodo (newTodo) {
+  const [todoList, setTodoList] = useState([]);
+  const [isLoading, setIsLoading] = useState(true);
+
+  const fetchData = () =>
+    new Promise((resolve, reject) =>
+      setTimeout(
+        () =>
+          resolve({
+            data: {
+              todoList: JSON.parse(localStorage.getItem("savedTodoList")) || [],
+            },
+          }),
+        2000
+      )
+    );
+
+  useEffect(() => {
+    fetchData().then((result) => {
+      setTodoList(result.data.todoList);
+      setIsLoading(false);
+    });
+  }, []);
```

Comment:
Recommended changes (not required):
This `fetchData` function should be defined within the `useEffect` function for 2 reasons:
1. It's only being utilized within this useEffect
2. This useEffect requires that the `fetchData` function be a part of its dependency array:
```js
  useEffect(() => {
    fetchData().then((result) => {
      setTodoList(result.data.todoList);
      setIsLoading(false);
    });
  }, [fetchData]);
```
however, when adding functions that are defined within a component body (above the return statement) to a dependency array must be wrapped in a [useCallback](https://react.dev/reference/react/useCallback) function. Rather than going down this more complex route, I'd highly recommend moving the `new Promise` into the useEffect:
```js
 useEffect(() => {
   new Promise((resolve, reject) =>
      setTimeout(
        () =>
          resolve({
            data: {
              todoList: JSON.parse(localStorage.getItem("savedTodoList")) || [],
            },
          }),
        2000
      )
    ).then((result) => {
      setTodoList(result.data.todoList);
      setIsLoading(false);
    });
  }, []);
```